genie-web/src/main/java/com/netflix/genie/web/services/impl/JobResolverServiceImpl.java

```diff
@@ -968,4 +1034,63 @@ private String mergeCriteriaStrings(
             throw new IllegalArgumentException(fieldName + "'s were both present but not equal");
         }
     }
+
+    /**
+     * This is an in memory evaluation of the matching done against persistence.
+     *
+     * @param cluster   The cluster to evaluate the criterion against
+     * @param criterion The criterion the cluster is being tested against
+     * @return {@literal true} if the {@link Cluster} matches the {@link Criterion}
+     */
+    private boolean clusterMatchesCriterion(final Cluster cluster, final Criterion criterion) {
+        // TODO: This runs the risk of diverging from DB query mechanism. Perhaps way to unite somewhat?
+        final ClusterMetadata metadata = cluster.getMetadata();
+
+        return criterion.getId().map(id -> cluster.getId().equals(id)).orElse(true)
+            && criterion.getName().map(name -> metadata.getName().equals(name)).orElse(true)
+            && criterion.getVersion().map(version -> metadata.getVersion().equals(version)).orElse(true)
+            && criterion.getStatus().map(status -> metadata.getStatus().name().equals(status)).orElse(true)
+            && metadata.getTags().containsAll(criterion.getTags());
+    }
+
+    private Map<Command, Set<Cluster>> generateCommandClustersMap(
+        final Map<Command, List<Criterion>> commandClusterCriteria,
+        final Set<Cluster> candidateClusters
+    ) {
+        final ImmutableMap.Builder<Command, Set<Cluster>> matrixBuilder = ImmutableMap.builder();
+        for (final Map.Entry<Command, List<Criterion>> entry : commandClusterCriteria.entrySet()) {
+            final Command command = entry.getKey();
+            final ImmutableSet.Builder<Cluster> matchedClustersBuilder = ImmutableSet.builder();
+
+            // Loop through the criterion in the priority order first
+            for (final Criterion criterion : entry.getValue()) {
+                for (final Cluster candidateCluster : candidateClusters) {
+                    if (this.clusterMatchesCriterion(candidateCluster, criterion)) {
+                        log.debug(
+                            "Cluster {} matched criterion {} for command {}",
+                            candidateCluster.getId(),
+                            criterion,
+                            command.getId()
+                        );
+                        matchedClustersBuilder.add(candidateCluster);
+                    }
+                }
+
+                final ImmutableSet<Cluster> matchedClusters = matchedClustersBuilder.build();
+                if (!matchedClusters.isEmpty()) {
+                    // If we found some clusters the evaluation for this command is done
+                    matrixBuilder.put(command, matchedClusters);
+                    log.debug("For command {} matched clusters {}", command, matchedClusters);
+                    // short circuit further criteria evaluation for this command
```

Comment:
Seems like this short-circuiting could create a matrix where only a subset of the potential matches for a commands are present. Is this ok?

Wouldn't the lack of some clusters (that actually ARE a match, but not present in this set) influence the command selector?

Reply:
This is a legit question. Basically you're asking if the matrix should instead be `Command -> List<Set<cluster>>` where we fully explore the combinations. As of now this logic matches what v3 did in terms of short circuiting the evaluation at the highest priority combination found to have matches. The selectors for clusters are guaranteed to select at least one if the random selector remains the last fallback.

The question here is if the command selector will make different decision based on presence of more clusters at a lower priority. Perhaps. @rdblue or Dan (couldn't find his git handle) may have an opinion however the only way I see this working is if the `CommandSelector` and `ClusterSelector` are combined as there (currently at least) is no way for the command selector to pass back information like "ok use this command but use the 3rd highest priority set of clusters not the first". We'd wnat it to select the command and the cluster.

For now I think this is something we could modify in the future if desired unless Ryan or Dan believe it critical to continuing with their work.

Reply:
I think this is correct, and a good way to avoid computing the entire intersection. We only need the highest priority match, just like before.